### PR TITLE
feat : dispute function

### LINKED ISSUE #151: Implement Admin Function to Resolve Disputes
## 📚 Overview

Implement a new function that allows the admin to resolve disputes raised on a pool. Once the dispute is resolved, the pool status should be updated accordingly, allowing reward claims and validation actions to resume.

## 🚀 Background

In the PrediFi platform, disputes should be raised by users or validators when pool results are suspected to be erroneous or manipulated. When a dispute is raised, the pool status should change to suspended and all actions (e.g., claiming rewards, validating outcomes) should be disabled also. The admin or DAO then steps in to review and resolve the dispute.

## 🛠️ Requirements

1. Create a function called `raise_dispute`
    - The function checks the number of people that have raised dispute, when the users that have raised dispute are more than 3 then the pool shoudl be suspended, actions like validate and claim reward should not work once a pool is suspended
2. Create a function named `resolve_dispute` (or follow the project's naming convention) for the admin/DAO to resolve a raised dispute.
   - The function should change the dispute status of the pool to what was formerly there.
   - Upon resolution, the pool should resume normal operations (reward claims, validations, etc.).
3. Create neccessary read functionf for it

4. Technical Considerations:
   - Ensure that dispute resolution can only be executed by authorized users (admin).
   - Validate that the pool is in a disputed state before allowing resolution.
   - Update any state variables associated with the pool's status to reflect that the dispute has been resolved.

## 🎯 Goals

- Maintain the security and trust of the platform by ensuring only authorized parties can resolve disputes.
- Seamlessly transition the pool from a suspended state back to normal operations once the dispute is resolved.

## 📑 Acceptance Criteria

- The function `create_dispute` is implemented and called by any contract address
- The function `resolve_dispute` is implemented and callable by the admin/DAO only.
- The pool state is correctly updated upon dispute resolution.
- Normal operations (reward claims and validations) resume once the dispute is resolved.
- All necessary tests pass locally and any new test cases cover this functionality.

Feel free to add any additional test parameters and documentation updates as needed. For more details on project setup, refer to the

### PR DESCRIPTION
**Pull request by JayWebtech**
Closes #151

### CODE REVIEW

**Comment by Akshola00** [comment]
gm @JayWebtech 
ci is failing

**Comment by JayWebtech** [comment]
image

@Akshola00 seems it works now. issue was related to scarb fmt

File: src/base/errors.cairo [hunk 1]
```diff
@@ -11,4 +11,10 @@ pub mod Errors {
     pub const TOTAL_STAKE_MUST_BE_ONE_STRK: felt252 = 'Total stake should be 1 STRK';
     pub const TOTAL_SHARE_MUST_BE_ONE_STRK: felt252 = 'Total shares should be 1 STRK';
     pub const USER_SHARE_MUST_BE_ONE_STRK: felt252 = 'User shares should be 1 STRK';
+    pub const POOL_SUSPENDED: felt252 = 'Pool is suspended';
```
Comment (Akshola00) [inline src/base/errors.cairo:14]: nice detailed error messages for edge cases

File: src/interfaces/ipredifi.cairo [hunk 2]
```diff
@@ -78,4 +78,15 @@ pub trait IPredifi<TContractState> {
     fn get_locked_pools(self: @TContractState) -> Array<PoolDetails>;
     fn get_settled_pools(self: @TContractState) -> Array<PoolDetails>;
     fn get_closed_pools(self: @TContractState) -> Array<PoolDetails>;
+
+    //dispute functionality
+    fn raise_dispute(ref self: TContractState, pool_id: u256);
```
Comment (Akshola00) [inline src/interfaces/ipredifi.cairo:83]: great traits for both getters and setter functions including helper functions

File: src/predifi.cairo [hunk 3]
```diff
@@ -93,6 +96,11 @@ pub mod Predifi {
         >, // Tracks how many pool IDs are stored for each user
         // Mapping to track which validators are assigned to which pools
         pool_validator_assignments: Map<u256, (ContractAddress, ContractAddress)>,
+        // Dispute functionality storage
+        pool_dispute_users: Map<(u256, ContractAddress), bool>,
```
Comment (Akshola00) [inline src/predifi.cairo:100]: storing relevant data in state, lgtm

File: src/predifi.cairo [hunk 4]
```diff
@@ -179,6 +190,26 @@ pub mod Predifi {
         pub caller: ContractAddress,
     }
 
+    #[derive(Drop, starknet::Event)]
```
Comment (Akshola00) [inline src/predifi.cairo:193]: neccessary events to be emmited after state changes ✅

File: src/predifi.cairo [hunk 5]
```diff
@@ -798,6 +833,130 @@ pub mod Predifi {
         fn get_closed_pools(self: @ContractState) -> Array<PoolDetails> {
             self.get_pools_by_status(Status::Closed)
         }
+
+        // dispute functions
```
Comment (Akshola00) [inline src/predifi.cairo:837]: Great implementation of both getter and setter functions here ✅

File: tests/test_contract.cairo [hunk 6]
```diff
@@ -2928,3 +2928,538 @@ fn test_get_closed_pools() {
...
+// DISPUTE FUNCTIONALITY TESTS
+// ========================================
+
+// Helper function to create a user address
+// fn create_user(index: felt252) -> ContractAddress {
+//     contract_address_const::<index>();
+// }
+
+// Helper function to approve and fund a user for testing
+fn setup_user_with_tokens(user: ContractAddress, erc20_address: ContractAddress, amount: u256) {
```
Comment (Akshola00) [inline tests/test_contract.cairo:2941]: detasiled tests covering state changes and edge cases here, sir ✅

**Review by Akshola00 — APPROVED** [review]
lgtm
